File: Day187_2025/Q187_GFG.java
import java.util.*;

class Solution {
    public ArrayList<Integer> topKSumPairs(int[] a, int[] b, int k) {
        int n = a.length;
        Arrays.sort(a);
        Arrays.sort(b);

        PriorityQueue<int[]> pq = new PriorityQueue<>((x, y) -> Integer.compare(y[0], x[0]));
        Set<String> visited = new HashSet<>();

        int i = n - 1, j = n - 1;
        pq.offer(new int[]{a[i] + b[j], i, j});
        visited.add(i + "," + j);

        ArrayList<Integer> result = new ArrayList<>();

        while (result.size() < k && !pq.isEmpty()) {
            int[] top = pq.poll();
            int sum = top[0], x = top[1], y = top[2];
            result.add(sum);

            if (x - 1 >= 0 && !visited.contains((x - 1) + "," + y)) {
                pq.offer(new int[]{a[x - 1] + b[y], x - 1, y});
                visited.add((x - 1) + "," + y);
            }

            if (y - 1 >= 0 && !visited.contains(x + "," + (y - 1))) {
                pq.offer(new int[]{a[x] + b[y - 1], x, y - 1});
                visited.add(x + "," + (y - 1));
            }
        }

        return result;
    }
}

File: Day187_2025/Q187_GFG.py
import heapq

class Solution:
    def topKSumPairs(self, a, b, k):
        n = len(a)
        a.sort(reverse=True)
        b.sort(reverse=True)
        
        visited = set()
        max_heap = []
        
        heapq.heappush(max_heap, (-(a[0] + b[0]), 0, 0))
        visited.add((0, 0))
        
        res = []
        while len(res) < k:
            curr_sum, i, j = heapq.heappop(max_heap)
            res.append(-curr_sum)
            
            if i + 1 < n and (i + 1, j) not in visited:
                heapq.heappush(max_heap, (-(a[i + 1] + b[j]), i + 1, j))
                visited.add((i + 1, j))
            
            if j + 1 < n and (i, j + 1) not in visited:
                heapq.heappush(max_heap, (-(a[i] + b[j + 1]), i, j + 1))
                visited.add((i, j + 1))
        
        return res
